### PR TITLE
Add Comment structure with links and tags

### PR DESCRIPTION
**Pull request by joffrey-bion**
This pull request implements a `CommentParser` responsible for creating a structure `Comment` object with pure `CommentText`s,  `InlineLink`s,  and more general `InlineTag`s.

The addition of `equals` and `hashcode` methods was only useful for writing tests more easily. They could be removed by changing how the tests are made, but I believe they don't hurt, and on the contrary they may be valuable later for clients of this lib too.

A next step would be to actually access imports, so that we could return `InlineLink`s with actual `Class` objects instead of class names.

Resolves:
https://github.com/dnault/therapi-runtime-javadoc/issues/17

### CODE REVIEW

File: therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    public void parse_mix_textAndTag() {
+        List<CommentElement> elements = CommentParser.parse("text before {@sometag some value}").getElements();
+        assertEquals(2, elements.size());
+        assertEquals(new CommentText("text before "), elements.get(0));
+        assertEquals(new InlineTag("sometag", "some value"), elements.get(1));
+    }
+
+    @Test
+    public void parse_mix_textAndLink() {
+        List<CommentElement> elements = CommentParser.parse("text before {@link ClassName} text after").getElements();
```
Comment (ST-DDT) [inline therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java:123]: Please also add a test for a text like this:

`"Adds the given {@link Action}s to the queue."`

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    private static final Pattern inlineTag = Pattern.compile("\\{@(\\w+)\\s+(\\w[^}]+)\\}");
+
+    private static final Pattern inlineTagSplitter = Pattern.compile("((?=(\\{@))|(?<=}))");
+
+    private static final Pattern whitespace = Pattern.compile("\\s+");
+
+    private static final Pattern linkRefSplitter = Pattern.compile("#");
+
+    static Comment parse(String commentText) {
+        String[] parts = inlineTagSplitter.split(commentText);
```
Comment (ST-DDT) [inline therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java]: IMO this could be better solved with a inlineTag matcher find startIndex endIndex search. If you need an example I can write it later today.

**Review by ST-DDT — COMMENTED** [review]
How should we handle `{@literal }`?

File: therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java (hunk 1 above)
Comment (joffrey-bion) [reply therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java:123]: Isn't this what this test is about? Or do you mean with no whitespace following the closing brace?

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java (hunk 2 above)
Comment (joffrey-bion) [reply therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java]: Indeed this is a better approach, it was around 3-4am, probably the reason why I used lookaheads instead !

File: therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java (hunk 1 above)
Comment (ST-DDT) [reply therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java:123]: Yes. The main use cases usually dont break during a refactor, but these edge cases do.

**Comment by joffrey-bion** [comment]
I made a couple changes to improve this PR.

Lookaheads were overkill and added unnecessary complexity that required clean up. Things like "foo}bar" were broken down to `["foo}", "bar"]`, so we had to merge them as a single text part. Using a `Matcher` solved this problem the right way.

Empty/blank tags (like `{@tag}` or `{@tag   }` seem to be taken as tags by the Javadoc tool, so we should do so too. I added support for that.

Things like `foo {@ } bar` are forbidden by the Javadoc tool, so we should not worry too much about them. The `CommentParser` currently leaves them as plain text.

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java (hunk 2 above)
Comment (joffrey-bion) [reply therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java]: Done.

File: therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java (hunk 1 above)
Comment (joffrey-bion) [reply therapi-runtime-javadoc/src/test/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParserTest.java:123]: Done.

**Review by ST-DDT — APPROVED** [review]
I haven't tested it myself yet, but looks good to me.

**Comment by joffrey-bion** [comment]
A couple more additions:
- blank comments should not return a `Comment` object, because technically there is no comment. This would force the users to deal with the absence of comment themselves, which is not nice.
- comments should be trimmed. I can't think of a situation where people actually would want additional blank lines or space in their description. On the other hand, if the global parser for the javadoc block leaves whitespace when splitting blocks, we'd better make sure it is removed by the `CommentParser`.

For a more detailed list of use cases, please simply read the tests ;-)

**Comment by dnault** [comment]
Thanks, nice work!
